cli/src/repl.rs

```diff
@@ -1,47 +1,230 @@
+use rustyline::history::DefaultHistory;
+use rustyline::{
+    Cmd, ColorMode, DefaultEditor, Editor, Event, EventHandler, KeyCode, KeyEvent, Modifiers,
+};
+use std::collections::HashMap;
+use std::process::exit;
+
+type REPLEditor = Editor<(), DefaultHistory>;
+
+#[derive(Default)]
+struct REPLImporter<'a> {
+    #[allow(clippy::vec_box)]
+    // Box is used to ensure that the reference behind is still valid after vector's realloc
+    sources: Vec<Box<OwnedSource>>,
+    imported_modules: HashMap<Name, Expr<'a>>,
+}
+
+impl<'a, const N: usize> From<[(Name, Expr<'a>); N]> for REPLImporter<'a> {
+    fn from(value: [(Name, Expr<'a>); N]) -> Self {
+        Self {
+            imported_modules: HashMap::from(value),
+            sources: Vec::new(),
+        }
+    }
+}
+
+impl<'a> ASTImporter<'a> for REPLImporter<'a> {
+    fn import(&mut self, name: &Name) -> Option<Expr<'a>> {
+        self.imported_modules.get(name).cloned()
+    }
+}
+
+impl<'a> REPLImporter<'a> {
+    pub fn take_source(&mut self, source: OwnedSource) -> Source<'a> {
+        self.sources.push(Box::new(source));
+        let src = self.sources[self.sources.len() - 1].as_source();
+        unsafe {
+            // SAFETY: The sources will never be removed from the self.sources vector as the REPLImporter's
+            // role is to be the owner of the user's sources.
+            // The reference behind Box does not change and is valid for the lifetime of the importer.
+            std::mem::transmute::<Source, Source<'a>>(src)
+        }
+    }
+}
 
 /// Indefinitely prompts a new expression to the stdin,
 /// displaying back the errors if any and the formed AST
-pub fn prompt() {
-    while let Some(source) = parse_input() {
-        handle_source(source.as_source());
+pub fn repl(config: Configuration) {
+    let mut editor: REPLEditor =
+        DefaultEditor::new().expect("unable to instantiate terminal editor");
+    editor.set_color_mode(ColorMode::Enabled);
+    editor.set_history_ignore_dups(true).unwrap();
+    editor.set_history_ignore_space(true);
+    editor.bind_sequence(
+        Event::KeySeq(vec![KeyEvent(KeyCode::Char('u'), Modifiers::ALT)]),
+        EventHandler::from(Cmd::Undo(1)),
+    );
+
+    let mut relations = Relations::default();
+    let mut imports = Imports::default();
+    let mut importer = REPLImporter::default();
+
+    loop {
+        let mut engine = Engine::default();
+
+        let source = parse_input(&mut editor);
+        handle_source(
+            source,
+            &config,
+            &mut engine,
+            &mut importer,
+            &mut imports,
+            &mut relations,
+        );
+    }
+}
+
+fn strip_indent(line: &mut String) -> String {
+    let mut indent = String::new();
+    let line_clone = line.to_string();
+    let chars = line_clone.chars();
+    for c in chars {
+        if !c.is_whitespace() {
+            return indent;
+        }
+        line.remove(0);
+        indent.push(c);
     }
+    indent
 }
```

Comment:
```suggestion
```
Stripping whitespaces is the job of the parser.

Reply:
This is meant to keep the same indent for multi line prompts and keep an history without the indent